src/pip/_internal/network/auth.py

```diff
@@ -170,13 +170,12 @@ def _get_url_and_credentials(self, original_url):
         """
         url, netloc, _ = split_auth_netloc_from_url(original_url)
 
-        # Use any stored credentials that we have for this netloc
-        username, password = self.passwords.get(netloc, (None, None))
+        # Try to get credentials from original url
+        username, password = self._get_new_credentials(original_url)
 
+        # If credentials not found, use any stored credentials for this netloc
         if username is None and password is None:
-            # No stored credentials. Acquire new credentials without prompting
-            # the user. (e.g. from netrc, keyring, or the URL itself)
-            username, password = self._get_new_credentials(original_url)
+            username, password = self.passwords.get(netloc, (None, None))
```

Comment:
This new implementation seems different from the previous one IIRC? Your previous implementation would use the username/password from the URL if either field is empty in the cache entry, so if

* `self.passowords` have `("foo", None)`
* `self._get_new_credentials()` returns `(Anything, "bar")`

The end result would be `("foo", "bar")`.

But with this new implementation, the end result would be `("foo", None)` instead.

I’m not sure which behaviour is better, but we need to be able to decide on one and explain it to users.

Reply:
You're right, the behavior is different from the previous one.

First, I've noticed that `self.passwords` cannot contain None on only one side (username or password) according to the current implementation where `self.passwords` is updated.

So, from my understanding, the remaining choice to make is when : 
1/ self.passwords have `("someusername", "somepassword")`
2/ `self._get_new_credentials()` returns `("other", None)` : means only username is specified in the url `http://other@domain.com/path/` 

In that case, I've finally thought that it doesn't make sense to mix username from one source with password from other source. So I would prefer to keep `("other", "")` in the end.

Reply:
Thanks for the extra insight! That is reasonable to me, so the current implementation is “more correct” than the previous.